mesonbuild/compilers.py

```diff
@@ -447,22 +447,34 @@ def _should_prepend(cls, arg):
             return True
         return False
 
-    def to_native(self):
+    def _fix_broken_linkers(self):
         # Check if we need to add --start/end-group for circular dependencies
         # between static libraries.
-        if get_compiler_uses_gnuld(self.compiler):
-            group_started = False
-            for each in self:
-                if not each.startswith('-l') and not each.endswith('.a'):
-                    continue
-                i = self.index(each)
-                if not group_started:
-                    # First occurance of a library
-                    self.insert(i, '-Wl,--start-group')
-                    group_started = True
-            # Last occurance of a library
-            if group_started:
-                self.insert(i + 1, '-Wl,--end-group')
+        if not get_compiler_uses_gnuld(self.compiler):
+            return
+        group_started = False
+        # Must ensure that we don't accidentally nest --start/end-group
+        # since older `ld` versions (such as 2.20) barf on that:
+        # https://github.com/mesonbuild/meson/issues/1936
+        # We could add complicated logic here to detect nesting, but
+        # users should not be adding --start-group themselves. KISS for now.
+        for each in self:
+            if each.endswith('--start-group'):
+                return
+        for each in self:
+            if not each.startswith('-l') and not each.endswith('.a'):
+                continue
+            i = self.index(each)
```

Comment:
For a library at the end of the list, `self` is traversed 3 times. Are you expecting there to be duplicates? This could be `for i, each in enumerate(self):` if not.

If you're only inserting one start/end, then I think the previous loop could be integrated into this one as well.